go/streamlog/bind_variable.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package streamlog
+
+import (
+	"encoding/json"
+	"errors"
+
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+var ErrUnrecognizedBindVarType = errors.New("unrecognized bind variable type")
+
+// BindVariableValue is used to store querypb.BindVariable values.
+type BindVariableValue struct {
+	Type  string
+	Value []byte
+}
+
+// MarshalJSON renders the BindVariableValue as json and optionally redacts the value.
+func (bv BindVariableValue) MarshalJSON() ([]byte, error) {
+	out := map[string]interface{}{
+		"Type":  bv.Type,
+		"Value": bv.Value,
+	}
+	if GetRedactDebugUIQueries() {
+		out["Value"] = nil
+	}
+	return json.Marshal(out)
+}
+
+// BindVariable is a wrapper for marshal/unmarshaling querypb.BindVariable as json.
+// It ensures that the "Type" field is a string representation of the variable
+// type instead of an integer-based code that is less portable and human-readable.
+//
+// This allows a *querypb.BindVariable that would have marshaled
+// to this:
+//
+//	{"Type":10262,"Value":"FmtAtEq6S9Y="}
+//
+// to marshal to this:
+//
+//	{"Type":"VARBINARY","Value":"FmtAtEq6S9Y="}
+//
+// or if query redaction is enabled, like this:
+//
+//	{"Type":"VARBINARY","Value":null}
+type BindVariable struct {
+	Type   string
+	Value  []byte
+	Values []*BindVariableValue
+}
+
+// NewBindVariable returns a wrapped *querypb.BindVariable object.
+func NewBindVariable(bv *querypb.BindVariable) BindVariable {
+	newBv := BindVariable{
```

Comment:
We know the capacity of `newBv.Values` up front and should preallocate it here before the loop.